### DISCUSSION TITLE
[HARDWARE] Support for Apple Silicon and Intel Mac

### DISCUSSION THREAD

**Discussion by BlackCube008** [hlky/stable-diffusion-webui#210] [General]
Re-opening as this was closed a bit too early. I am still having troubles...
The first part was: https://github.com/hlky/stable-diffusion-webui/issues/145

Here my comments to yours
| conda env create -f environment.yaml

Hey thanks for this. I have some difficulties. Here is what I have done:

1. I have duplicated a file that ended with .sh
2. I have changed the content with the below and renamed the file webui.sh
3. I have placed that file on the location stable-diffusion-main
4. I have opened Terminal and have then written sh [space] then I have moved the filw webui.sh onto terminal to start the process
5. I have encountered an error

What am I doing wrong?
What should I write exactly after activate?

Thx!

---
conda env create -f /Users/[name_of_the_user]/Downloads/stable-diffusion-main/environment.yaml
conda env update --file environment.yaml --prune
conda activate /Users/[name_of_the_user]/Downloads/stable-diffusion-main/environment.yaml
python scripts\relauncher.py

**Comment by hlky** [comment]
What error do you get?

I'm not sure about mac but on linux which is similar you would probably need to do chmod +x 

> /Users/[name_of_the_user]/Downloads/stable-diffusion-main

This shouldn't be needed if the .sh script is in the stable-diffusion directory and the current working directory when you run the .sh script is the stable-diffusion folder

You may still have issues with some features, see https://github.com/magnusviri/stable-diffusion for Apple Silicon support.

Is your Mac actually Apple Silicon or is it an Intel Mac? Support for the two will probably differ.

**Comment by BlackCube008** [comment]
Hey,

I have an Intel and here is what I get by using the following content:

conda env create -f /Users/[name_of_the_account]/Downloads/stable-diffusion-main/environment.yaml
conda env update --file /Users/[name_of_the_account]/Downloads/stable-diffusion-main/environment.yaml --prune
conda activate /Users/[name_of_the_account]/Downloads/stable-diffusion-main/environment.yaml
python scripts\relauncher.py

---

(base) [name_of_the_account]@MacBook-Pro ~ % sh /Users/[name_of_the_account]/Downloads/stable-diffusion-main/webui.sh
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - cudatoolkit=11.3

Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - cudatoolkit=11.3


CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

IMPORTANT: You may need to close and restart your shell after running 'conda init'.


python: can't open file '/Users/[name_of_the_account]/scriptsrelauncher.py': [Errno 2] No such file or directory

**Comment by hlky** [comment]
Well do you even have a nvidia graphics card? Apple silicon is supported because it has good integrated graphics. Afaik Intel Mac have Intel Iris, unless you have a Mac Pro or something I think you're out of luck.

>/scriptsrelauncher.py

That specifically is because the slash is the wrong way round

**Reply by lwasyl** [reply]
What about a macbook pro with a dedicated AMD Radeon Pro Vega GPU, I suppose that's not supported either?

**Reply by magnusviri** [reply]
Someone reported [here](https://github.com/CompVis/stable-diffusion/pull/56#issuecomment-1239956456) that they got it working on a 2013 MacBook Pro.

**Reply by lwasyl** [reply]
I actually managed to run https://github.com/lstein/stable-diffusion with a 2018 Macbook Pro with Radeon Vega 20 with ~5s/iter by following the instructions there. It would be great to use this webui though, as it's much nicer than the one at @\lstein

**Reply by lwasyl** [reply]
So it looks like intel/m1 macbooks could be supported by using pytorch's `mps` device which supports both M1 and AMD GPUs. It doesn't seem like a lot of changes to support it: https://github.com/lstein/stable-diffusion/commit/bdb0651eb2120a5366dc64067add74fecf66db81 although it does require pytorch>=1.12 and I haven't been yet able to run mps on sd-webui repo myself. I'm not very familiar with python/pytorch/cuda so there's probably something simple I'm missing — right now I'm stuck on `view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.` error. 

Perhaps it'd be easier to add this support for someone more familiar with the scripts/ecosystem.

**Reply by lwasyl** [reply]
Also just a note — Intel Macs with Vega graphics perform very well, 3-5it/s, so it's not a completely unreasonable request to have MPS support

**Comment by BlackCube008** [comment]
I have this: Intel UHD Graphics 630 1536 MB
I have corrected but I still got an error.

can't open file '/Users/[name_of_the_account]/scripts/relauncher.py

	Relaunch count: 56
python: can't open file '/Users/[name_of_the_account]/scripts/webui.py': [Errno 2] No such file or directory but the file is there

**Comment by hlky** [comment]
`Intel UHD Graphics 630 1536 MB` a Mac with this configuration is unlikely to be supported except when #185 is added, or you try a different repo that has CPU support, although other repos won't have the same features as this one

`'/Users/[name_of_the_account]/scripts/webui.py':`
`/Users/[name_of_the_account]/Downloads/stable-diffusion-main/`
wrong path and 

> This shouldn't be needed if the .sh script is in the stable-diffusion directory and the current working directory when you run the .sh script is the stable-diffusion folder

just `scripts/webui.py` or `scripts/relauncher.py` etc

Not that it matters as it's unlikely you can get this working with your hardware for now

**Reply by AK391** [reply]
this repo has cpu support with with OpenVINO, could add as a option in the gui?: https://github.com/bes-dev/stable_diffusion.openvino

**Reply by AK391** [reply]
also onnx weights: https://huggingface.co/bes-dev/stable-diffusion-v1-4-onnx

**Reply by hlky** [reply]
onnx weights? 

i've seen the openvino implementation, could be added for cpu support,  in regard to this issue it would only solve for Intel Mac, there's another repo with work done on Apple Silicon support

**Comment by vdruts** [comment]
Hey guys, I'm a total noob here and just wondering if there is a way to get it to work on M1 Mac especially with GUI yet?

**Reply by arbitraryvolume** [reply]
Would be great to be able to run this on M1 silicon GPU as pioneered by @magnusviri -- https://github.com/magnusviri/stable-diffusion >> and merged into https://github.com/lstein/stable-diffusion (https://github.com/lstein/stable-diffusion/blob/main/README-Mac-MPS.md) -- sadly stretches beyond my current skillset.

**Reply by edwios** [reply]
Yes, the Magnusviri fork works like a charm on the M1 and is really fast.
